### PR TITLE
AAD account checks on packages for safety reports

### PR DESCRIPTION
**Pull request by drewgillies**
We'll allow safety report categories only on MSA-only account-owned packages as a first step.

### CODE REVIEW

File: src/NuGetGallery.Core/CredentialTypes.cs [hunk 1]
```diff
@@ -114,6 +115,17 @@ public static bool IsAzureActiveDirectoryAccount(string type)
             return type?.Equals(External.AzureActiveDirectoryAccount, StringComparison.OrdinalIgnoreCase) ?? false;
         }
 
+        public static bool PackageHasNoAadOwners(Package package)
```
Comment (skofman1) [inline src/NuGetGallery.Core/CredentialTypes.cs:118]: Consider moving this check to the Controller. We can move to NuGetGallery.Core if we need to re-use it.

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 2]
```diff
@@ -1384,7 +1384,8 @@ public virtual ActionResult ReportAbuse(string id, string version)
 
             var model = new ReportAbuseViewModel
             {
-                ReasonChoices = _featureFlagService.IsShowReportAbuseSafetyChangesEnabled()
+                ReasonChoices = _featureFlagService.IsShowReportAbuseSafetyChangesEnabled() 
+                    && CredentialTypes.PackageHasNoAadOwners(package)
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/PackagesController.cs:1388]: UTs?

File: src/NuGetGallery.Core/CredentialTypes.cs [hunk 3]
```diff
@@ -114,6 +115,17 @@ public static bool IsAzureActiveDirectoryAccount(string type)
...
+        {
+            var owners = package?.PackageRegistration?.Owners;
+            if (owners == null || !owners.Any()) 
+            {
+                return true;
+            }
+
+            return !owners.Where(o => o.Credentials.GetAzureActiveDirectoryCredential() != null).Any();
```
Comment (skofman1) [inline src/NuGetGallery.Core/CredentialTypes.cs:126]: Will !owners.Any(o => o.Credentials.GetAzureActiveDirectoryCredential() != null);
work?

Comment (camigthompson) [reply src/NuGetGallery.Core/CredentialTypes.cs:126]: Not sure if this is valid or not, but could you do `return !package?.PackageRegistration?.Owners?.Any(o => o.Credentials.GetAzureActiveDirectoryCredential() != null) ?? true`? Just to mimic the other methods here.

Comment (drewgillies) [reply src/NuGetGallery.Core/CredentialTypes.cs]: @skofman1 yes, that's cleaner--thanks. @camigthompson I'd like to leave it as two statements for readability, particularly as I'm separating our NRE-avoidance into its own statement.

**Comment by skofman1** [comment]
What should be the behavior if one of the package owners is an organization, and the org has collaborators that use AAD?

**Comment by drewgillies** [comment]
@skofman1 This case is addressed now. Thanks for pointing it out!

**Review by joelverhagen — APPROVED** [review]

**Pull request merged**